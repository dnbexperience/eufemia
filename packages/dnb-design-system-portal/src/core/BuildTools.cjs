const path = require('path')
const fs = require('fs-extra')

function shouldUsePrebuild() {
  const rootPath = path.dirname(require.resolve('@dnb/eufemia'))
  const buildPath = path.resolve(rootPath, 'build')

  global.PREBUILD_EXISTS = fs.existsSync(buildPath)

  return Boolean(
    global.PREBUILD_EXISTS && process.env.NODE_ENV === 'production'
  )
}
module.exports.shouldUsePrebuild = shouldUsePrebuild
