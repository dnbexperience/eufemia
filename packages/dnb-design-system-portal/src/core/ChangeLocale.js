import React from 'react'
// import { ToggleButton } from '@dnb/eufemia/src'
import { Dropdown } from '@dnb/eufemia/src'

import Context from '@dnb/eufemia/src/shared/Context'
import { setLang } from './portalProviders'

export default function ChangeLocale({ showUS, ...props }) {
  const { locale, setLocale } = React.useContext(Context)

  // return (
  //   <ToggleButton.Group
  //     value={locale}
  //     on_change={({ value }) => {
  //       setLocale(value)
  //       setLang(value)
  //     }}
  //     {...props}
  //   >
  //     <ToggleButton value="nb-NO">Norsk</ToggleButton>
  //     <ToggleButton value="en-GB">English</ToggleButton>
  //     <ToggleButton value="en-US">English (US)</ToggleButton>
  //     {/* <ToggleButton value="de-DE">DE</ToggleButton> */}
  //   </ToggleButton.Group>
  // )

  const date = {
    'nb-NO': 'Norsk',
    'en-GB': 'English (GB)',
  }
  if (showUS) {
    date['en-US'] = 'English (US)'
  }

  return (
    <Dropdown
      value={locale}
      data={date}
      on_change={({ data: { value } }) => {
        setLocale(value)
        setLang(value)
      }}
      {...props}
    />
  )
}
