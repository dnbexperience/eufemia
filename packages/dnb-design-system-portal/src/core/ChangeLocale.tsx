import React from 'react'
import Context, { InternalLocale } from '@dnb/eufemia/src/shared/Context'
import { setLang } from './PortalProviders'
import { Field } from '@dnb/eufemia/src/extensions/forms'

export default function ChangeLocale({ listUSLocale = null, ...props }) {
  const { locale, setLocale } = React.useContext(Context)

  return (
    <Field.Selection
      value={locale}
      onChange={(value) => {
        setLocale(value as InternalLocale)
        setLang(value)
      }}
      {...props}
    >
      <Field.Option value="nb-NO" title="Norsk" />
      <Field.Option value="en-GB" title="English (GB)" />
      {listUSLocale && <Field.Option value="en-US" title="English (US)" />}
    </Field.Selection>
  )
}
