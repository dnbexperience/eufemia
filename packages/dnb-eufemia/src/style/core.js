/**
 * Imports only the core styles
 *
 */

import './dnb-ui-core.scss'
