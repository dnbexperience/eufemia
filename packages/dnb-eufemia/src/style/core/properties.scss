/*
 * CSS Properties
 *
 */

@mixin getDefaultProperties() {
  // Typography Family
  --font-family-default: 'DNB', sans-serif;
  --font-family-monospace: 'DNBMono', 'Menlo', 'Consolas', 'Roboto Mono',
    'Ubuntu Monospace', 'Noto Mono', 'Oxygen Mono', 'Liberation Mono',
    monospace;

  // Deprecated
  --font-family-book: 'DNB', sans-serif; // Deprecated
  --font-family-demi: 'DNB', sans-serif; // Deprecated

  // Typography Weights
  --font-weight-basis: normal;
  --font-weight-regular: normal;
  --font-weight-medium: 500;
  --font-weight-bold: 600;

  // Deprecated
  --font-weight-default: normal; // Deprecated
  --font-weight-book: normal; // Deprecated
  --font-weight-demi: 500; // Deprecated

  // Typography Sizes
  --font-size-x-small: 0.875rem; // 14px
  --font-size-small: 1rem; // 16px
  --font-size-basis: 1.125rem; // 18px
  --font-size-basis--em: 1em; // 16px
  --font-size-medium: 1.25rem; // 20px
  --font-size-large: 1.625rem; // 26px
  --font-size-x-large: 2.125rem; // 34px
  --font-size-xx-large: 3rem; // 48px

  // Typography Line heights
  --line-height-xx-small--em: 1em; // 16px
  --line-height-x-small: 1.125rem; // 18px
  --line-height-small: 1.25rem; // 20px
  --line-height-basis: 1.5rem; // 24px
  --line-height-basis--em: 1.333em; // 1.33333333333*18 = 24
  --line-height-medium: 2rem; // 32px
  --line-height-large: 2.5rem; // 40px
  --line-height-x-large: 3.5rem; // 56px

  // UX colors
  --color-mint-green-50: #d2f0e9;
  --color-mint-green-25: #e9f8f4;
  --color-mint-green-12: #f4fbf9;
  --color-sea-green-30: #b3dada;
  --color-sea-green-alt-30: #b3dada; // Deprecated
  --color-signal-yellow-30: #ffffd7; // Deprecated
  --color-accent-yellow-30: #feebc1;
  --color-signal-orange: #ff5400;
  --color-fire-red: #dc2a2a;
  --color-success-green: #008000;
  --color-fire-red-8: #fdeeee;
  --color-black: #000;
  --color-black-80: #333;
  --color-black-55: #737373;
  --color-black-30: #b3b3b3; // Deprecated
  --color-black-20: #ccc;
  --color-black-8: #ebebeb;
  --color-black-3: #f8f8f8;
  --color-white: #fff;

  // Undefined
  --color-black-border: #cdcdcd;
  --color-black-background: #fafafa;

  // Profile colors
  --color-sea-green: #007272;
  --color-sea-green-alt: #007272; // Deprecated
  --color-mint-green: #a5e1d2;
  --color-summer-green: #28b482;
  --color-emerald-green: #14555a;
  --color-ocean-green: #00343e;
  --color-signal-yellow: #ffff7a; // Deprecated
  --color-accent-yellow: #fdbb31;
  --color-indigo: #23195a;
  --color-violet: #6e2382;
  --color-sky-blue: #4bbed2;
  --color-lavender: #f2f2f5;
  --color-sand-yellow: #fbf6ec;
  --color-pistachio: #f2f4ec;

  // Unused Profile colors
  --color-mint-green-alt: #ebfffa;
  --color-indigo-medium: #6e6491;
  --color-indigo-light: #b9afc8;
  --color-violet-medium: #a06eaf;
  --color-violet-light: #cfb9d7;
  --color-sky-blue-medium: #87d2e1;
  --color-sky-blue-light: #c3ebf0;

  // Spacing
  --spacing-xx-small: 0.25rem;
  --spacing-x-small: 0.5rem;
  --spacing-small: 1rem;
  --spacing-medium: 1.5rem;
  --spacing-large: 2rem;
  --spacing-x-large: 3rem;
  --spacing-xx-large: 3.5rem;

  // Layout
  --layout-small: 40em; // 640
  --layout-medium: 50em; // 800
  --layout-large: 60em; // 960
  --layout-x-large: 72em; // 1152
  --layout-xx-large: 80em; // 1280
  --layout-xxx-large: 90em; // 1440
}

@mixin setProperties() {
  :root {
    @content;
  }
}

$hasCustomProperties: false !default;
@if $hasCustomProperties == false {
  :root {
    @include getDefaultProperties();
  }
}
