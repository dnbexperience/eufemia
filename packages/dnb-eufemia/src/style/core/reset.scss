/*
 * Document Reset
 *
 */

@mixin globalReset() {
  /**
   * 1. Use the default cursor in all browsers (opinionated).
   * 2. Use the default user interface font in all browsers (opinionated).
   * 3. Correct the line height in all browsers.
   * 4. Use a 4-space tab width in all browsers (opinionated).
   * 5. Prevent adjustments of font size after orientation changes in
   *    IE on Windows Phone and in iOS.
   * 6. Breaks words to prevent overflow in all browsers (opinionated).
   */

  html {
    cursor: default; /* 1 */
    font-family:
      system-ui,
      -apple-system,
      'Segoe UI',
      Roboto,
      Ubuntu,
      Cantarell,
      'Noto Sans',
      sans-serif,
      'Apple Color Emoji',
      'Segoe UI Emoji',
      'Segoe UI Symbol',
      'Noto Color Emoji'; /* 2 */

    line-height: var(--line-height-basis); /* 3 */ /* 4 */
    tab-size: 4; /* 4 */ /* 5 */
    text-size-adjust: 100%; /* 5 */
    word-break: break-word; /* 6 */
  }

  /**
   * Remove the margin in all browsers (opinionated).
   */

  body {
    margin: 0;
  }
}

@mixin coreReset() {
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */

  *,
  ::before,
  ::after {
    background-repeat: no-repeat; /* 1 */
    box-sizing: border-box; /* 2 */
  }

  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */

  ::before,
  ::after {
    text-decoration: inherit; /* 1 */
    vertical-align: inherit; /* 2 */
  }
}

@mixin reset() {
  @include coreReset();

  /* Grouping content
   */

  /**
   * 1. Add the correct sizing in Firefox.
   * 2. Show the overflow in Edge and IE.
   */

  hr {
    height: 0; /* 1 */
    overflow: visible; /* 2 */
  }

  /**
   * Add the correct display in IE.
   */

  main {
    display: block;
  }

  /**
   * Remove the list style on navigation lists in all browsers (opinionated).
   */

  nav ol,
  nav ul {
    list-style: none;
  }

  /* Text-level semantics
   */

  /**
   * Add the correct text decoration in Edge, IE, Opera, and Safari.
   */

  abbr[title] {
    text-decoration: underline;
    text-decoration: underline dotted;
  }

  /**
   * Add the correct font weight in Chrome, Edge, and Safari.
   */

  b,
  strong {
    font-weight: bolder;
  }

  /**
   * 1. Use the default monospace user interface font
   *    in all browsers (opinionated).
   */

  pre,
  code,
  kbd,
  samp {
    // Why do we do this in a reset?
    font-family: var(--font-family-monospace); /* 1 */
  }

  /*
   * Remove the text shadow on text selections in Firefox 61- (opinionated).
   * 1. Restore the coloring undone by defining the text shadow
   *    in all browsers (opinionated).
   */

  ::selection {
    background-color: #b3d4fc; /* 1 */
    color: #000; /* 1 */
    text-shadow: none;
  }

  /* Embedded content
   */

  /*
   * Change the alignment on media elements in all browsers (opinionated).
   */

  audio,
  canvas,
  iframe,
  img,
  svg,
  video {
    vertical-align: middle;
  }

  /**
   * Add the correct display in IE 9-.
   */

  audio,
  video {
    display: inline-block;
  }

  /**
   * Add the correct display in iOS 4-7.
   */

  audio:not([controls]) {
    display: none;
    height: 0;
  }

  /**
   * Remove the border on images inside links in IE 10-.
   */

  img {
    border-style: none;
  }

  /**
   * Change the fill color to match the text color in all browsers (opinionated).
   */

  svg:not([fill]) {
    fill: currentcolor;
  }

  /**
   * Hide the overflow in IE.
   */

  svg:not(:root) {
    overflow: hidden;
  }

  /* Tabular data
   */

  /**
   * Collapse border spacing in all browsers (opinionated).
   */

  table {
    border-collapse: collapse;
  }

  /* Forms
   */

  /**
   * Inherit styling in all browsers (opinionated).
   */

  button,
  input,
  select,
  textarea {
    font-family: inherit;
    font-size: inherit;
    line-height: inherit;
  }

  /**
   * Remove the margin in Safari.
   */

  figure,
  button,
  input,
  select {
    margin: 0;
  }

  /**
   * 1. Show the overflow in IE.
   * 2. Remove the inheritance of text transform in Edge, Firefox, and IE.
   */

  button {
    overflow: visible; /* 1 */
    text-transform: none; /* 2 */
  }

  /**
   * Correct the padding in Firefox.
   */

  fieldset {
    padding: 0.35rem 0.75rem 0.625rem;
  }

  /**
   * Show the overflow in Edge and IE.
   * 1. Safari did inherit the break-word from html root
   */

  input {
    overflow: visible;
    word-break: normal; /* 1 */
  }

  /**
   * 1. Correct the text wrapping in Edge and IE.
   * 2. Correct the color inheritance from `fieldset` elements in IE.
   */

  legend {
    color: inherit; /* 2 */
    display: table; /* 1 */
    max-width: 100%; /* 1 */
    white-space: normal; /* 1 */
  }

  /**
   * 1. Add the correct display in Edge and IE.
   * 2. Add the correct vertical alignment in Chrome, Firefox, and Opera.
   */

  progress {
    display: inline-block; /* 1 */
    vertical-align: baseline; /* 2 */
  }

  /**
   * Remove the inheritance of text transform in Firefox.
   */

  select {
    text-transform: none;
  }

  /**
   * 2. Remove the default vertical scrollbar in IE.
   * 3. Change the resize direction on textareas in all browsers (opinionated).
   */

  textarea {
    overflow: auto; /* 2 */
    resize: vertical; /* 3 */
  }

  /**
   * Remove the padding in IE 10-.
   */

  [type='checkbox'],
  [type='radio'] {
    padding: 0;
  }

  /**
   * 1. Correct the odd appearance in Chrome and Safari.
   * 2. Correct the outline style in Safari.
   */

  input[type='search'] {
    appearance: textfield; /* 1 */
    outline-offset: -2px; /* 2 */
  }

  /**
   * Correct the cursor style of increment and decrement buttons in Safari.
   */

  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    height: auto;
  }

  /**
   * Correct the text style of placeholders in Chrome, Edge, and Safari.
   */

  // stylelint-disable-next-line
  ::input-placeholder {
    color: inherit;
    opacity: 0.54;
  }

  /**
   * Remove the inner padding in Chrome and Safari on macOS.
   */

  ::-webkit-search-decoration {
    appearance: none;
  }

  /**
   * 1. Change font properties to `inherit` in Safari.
   */

  ::-webkit-file-upload-button {
    font: inherit; /* 1 */
    background-color: transparent;
    border: none;
  }

  /**
   * Remove the inner border and padding of focus outlines in Firefox.
   */

  ::-moz-focus-inner {
    border-style: none;
    padding: 0;
  }

  /* Interactive
   */

  /*
   * Add the correct display in Edge and IE.
   */

  details {
    display: block;
  }

  /*
   * Add the correct display in all browsers.
   */

  summary {
    display: list-item;
  }

  /* Scripting
   */

  /**
   * Add the correct display in IE 9-.
   */

  canvas {
    display: inline-block;
  }

  /**
   * Add the correct display in IE.
   */

  template {
    display: none;
  }

  /* User interaction
   */

  /*
   * 1. Remove the tapping delay in IE 10.
   * 2. Remove the tapping delay on clickable elements
        in all browsers (opinionated).
   */

  a,
  area,
  button,
  input,
  label,
  select,
  summary,
  textarea,
  [tabindex] {
    /* 1 */
    touch-action: manipulation; /* 2 */
  }

  /**
   * Add the correct display in IE 10-.
   */

  [hidden] {
    display: none;
  }

  /* Accessibility
   */

  /**
   * Change the cursor on busy elements in all browsers (opinionated).
   */

  [aria-busy='true'] {
    cursor: progress;
  }

  /*
   * Change the cursor on control elements in all browsers (opinionated).
   */

  [aria-controls]:not(input) {
    cursor: pointer;
  }

  /*
   * Change the cursor on disabled, not-editable, or otherwise
   * inoperable elements in all browsers (opinionated).
   */

  [aria-disabled='true'],
  [disabled] {
    cursor: not-allowed;
  }

  /*
   * Change the display on visually hidden accessible elements
   * in all browsers (opinionated).
   */

  [aria-hidden='false'][hidden]:not(:focus) {
    clip: rect(0, 0, 0, 0);
    display: inherit;
    position: absolute;
  }
}

// reset ePlatform css
@mixin resetLegacyStyles() {
  html[xmlns="http://www.w3.org/1999/xhtml"]
  {
    // reset underline during hover
    a:hover {
      text-decoration: none;
    }

    .dnb-anchor--active {
      color: var(--color-mint-green) !important;
    }

    .dnb-anchor--contrast:not(:hover) {
      color: var(--color-white) !important;
    }

    a.dnb-button--primary {
      color: var(--color-white);
    }

    // reset lists
    ul {
      list-style: initial;
    }

    // reset paragraphs
    p {
      overflow: initial;
      text-align: inherit;
      background-color: initial;
    }

    i {
      font-style: italic;
    }

    // make content responsive, also for 200% in font-size
    #column_left,
    #root {
      width: 100%;
    }

    body#dnbLayoutDefault #wrapper,
    body#dnbLayoutDefaultStartPage #wrapper {
      width: 100%;
      max-width: 60rem;
    }

    // reset the main content shadow
    #column_content {
      box-shadow: none;
    }
  }
}
