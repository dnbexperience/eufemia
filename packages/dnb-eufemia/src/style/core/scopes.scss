/*
 * Scopes
 *
 */

@import './reset.scss';

@mixin coreDefault() {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: var(--font-size-small); // has to be 16px
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);

  // The new DNB font needs font smoothing to be thinner on webkit and FF
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;

  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);

  /**
  * Ensure consistency and use the same as HTML reset -> html {...}
  * between base and code package
  */
  -moz-tab-size: 4;
  tab-size: 4;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  word-break: break-word;

  @content;
}

@mixin bodyDefault() {
  @include reset();
  @include coreDefault();

  @media print {
    // Use hard coded colors, because of theming and color swaps
    background-color: white;
    color: black;
  }

  @content;
}

@mixin htmlDefault() {
  html {
    scroll-behavior: smooth;
    font-size: 100%;

    // IS_SAFARI_MOBILE
    @supports (-webkit-touch-callout: none) {
      @supports (font: -apple-system-body) {
        font: -apple-system-body; /* stylelint-disable-line */
      }
    }
  }

  // reset ePlatform css
  @include resetLegacyStyles();
}

@mixin componentReset() {
  @include coreDefault();
  @include coreReset();

  margin: 0;
  padding: 0;

  @content;
}
