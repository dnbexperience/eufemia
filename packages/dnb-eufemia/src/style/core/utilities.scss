/*
 * Utilities
 */

// Mobile-first media query helper
@mixin media($bp) {
  @media screen and (min-width: #{$bp}) {
    @content;
  }
}

$defaultEasing: cubic-bezier(0.42, 0, 0, 1);
@mixin defaultDropShadow() {
  box-shadow: 0 1px 6px rgba(0, 0, 0, 0.16);
}

// Uses html:not([data-whatintent='touch'] )
// classes on <html> element
// based on js touch device test
@mixin hover() {
  html:not([data-whatintent='touch']) &:hover {
    &[disabled] {
      cursor: not-allowed;
    }
    &:not([disabled]) {
      @content;
    }
  }
}
@mixin active() {
  &:active,
  html:not([data-whatintent='touch']) &:active {
    &[disabled] {
      cursor: not-allowed;
    }
    // stylelint-disable-next-line
    &:not([disabled]) {
      @content;
    }
  }
}
@mixin focus() {
  &:focus,
  html:not([data-whatintent='touch']) &:focus {
    &[disabled] {
      cursor: not-allowed;
    }
    // stylelint-disable-next-line
    &:not([disabled]) {
      @content;
    }
  }
}

$focusRingWidth: 0.125rem; //2px
$focusRingColor: var(--color-emerald-green);
@mixin fakeFocus($whatinput: null, $color: null, $inset: null) {
  outline: none;

  @if $whatinput == null {
    $whatinput: 'keyboard';
  }
  @if $color == null {
    $color: $focusRingColor;
  }

  html[data-whatinput='#{$whatinput}'] & {
    @include fakeBorder($color, $focusRingWidth, $inset);
  }
}
@mixin removeFakeFocus($whatinput: null) {
  outline: initial;
  @if $whatinput == null {
    $whatinput: 'keyboard';
  }

  html[data-whatinput='#{$whatinput}'] & {
    box-shadow: none;
  }
}

@mixin fakeBorder(
  $color: null,
  $width: null /* 1px */,
  $inset: null,
  $is1px: false /* IE11 fix */,
  $important: null
) {
  @if $color == null {
    $color: $focusRingColor;
  }
  @if $width == null {
    $width: 0.0625rem;
  }

  --border-color: #{$color};

  // we use !important, cause we have no changes to select the selctor right
  // in some cases we have another state where we use box-shadow
  // but with important, we take care of that we actually can use use it
  box-shadow: $inset 0 0 0 $width var(--border-color) $important;

  @if $inset {
    /* iOS fix - because "inset" works not fine with border-radius */
    @include IS_SAFARI_MOBILE() {
      // remove inset on webkit - this makes the border outside
      box-shadow: 0 0 0 $width var(--border-color) $important;

      // This also works, but then we move content for 1px, if they don't have a border from before
      // box-shadow: none;
      // border: $width solid var(--border-color) $important;
    }

    /* Safari fix - because "inset" works not fine with border-radius if the user zooms the page */
    @include IS_SAFARI_DESKTOP() {
      // remove inset on webkit - this makes the border outside
      box-shadow: 0 0 0 $width var(--border-color) $important;
    }
  }

  @include IS_IE() {
    @if $is1px {
      // Since 0.0625rem gets not rendered correctly
      // and gives an artifact, we do use px instead
      box-shadow: $inset 0 0 0 1px $color $important;
    } @else {
      // IE color fallback
      box-shadow: $inset 0 0 0 $width $color $important;
    }
  }

  @if $color != null {
    border-color: transparent;
  }
}

@mixin extendFakeFocus(
  $first-color: null,
  $second-color: null,
  $width: 0.0625rem /*1px*/
) {
  $second: 0 0 0 ($focusRingWidth + $width) $second-color;
  box-shadow: 0 0 0 $width $first-color, $second;

  @include IS_IE() {
    border: 0 solid transparent; // only to trigger the renderering correctly in IE11
  }
}

@mixin dummySpacing() {
  // we use "aria-hidden" SPAN to simulate a wider width for each tab
  .dnb-dummy {
    display: flex;
    flex-direction: column;

    height: 0;
    visibility: hidden;
    overflow: hidden;
  }
}

@mixin scrollY($mode: scroll) {
  overflow-y: $mode;
  overscroll-behavior: contain;
  scrollbar-width: thin;
  html:not([data-visual-test]) & {
    scroll-behavior: smooth;
  }

  // older iOS safari
  -webkit-overflow-scrolling: touch;

  // show scrollbar on IE
  -ms-overflow-style: auto;
}

@mixin scrollX($mode: scroll) {
  overflow-x: $mode;
  overscroll-behavior: contain;
  scrollbar-width: thin;
  scroll-behavior: smooth;

  html[data-visual-test] & {
    scroll-behavior: auto !important;
  }

  // older iOS safari
  -webkit-overflow-scrolling: touch;

  // show scrollbar on IE
  -ms-overflow-style: auto;
}

@mixin hideScrollbar() {
  /* Hide scrollbar for Chrome, Safari */
  /* stylelint-disable-next-line */
  &::-webkit-scrollbar {
    display: none;
  }
  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */
}

// IE fix
@mixin IS_IE() {
  @media screen and (-ms-high-contrast: none) {
    @content;
  }
}

// Edge fix
@mixin IS_EDGE() {
  @supports (-ms-ime-align: auto) {
    @content;
  }
}

// Firefox fix
@mixin IS_FF() {
  @supports (-moz-appearance: meterbar) and
    (background-blend-mode: difference, normal) {
    @content;
  }
}

// Chrome fix
@mixin IS_CHROME() {
  @supports (-webkit-appearance: none) and
    (not (overflow: -webkit-marquee)) and (not (-ms-ime-align: auto)) and
    (not (-moz-appearance: none)) {
    @content;
  }
}

// Safari Mobile fix
@mixin IS_SAFARI_MOBILE() {
  @supports (-webkit-touch-callout: none) {
    @content;
  }
}

// Safari Desktop fix
@mixin IS_SAFARI_DESKTOP() {
  @media not all and (min-resolution: 0.001dpcm) {
    @supports (-webkit-appearance: none) and (stroke-color: transparent)
      and (not (-webkit-overflow-scrolling: touch)) {
      @content;
    }
  }
}

$breakpoints: (
  'small': 40em,
  'medium': 50em,
  'large': 60em,
  'x-large': 70em
);

//Example usage:
//@include allAbove(phone){ styles go here.. }
@mixin allAbove($size) {
  @media screen and (min-width: map-get($breakpoints, $size)) {
    @content;
  }
}
@mixin allBelow($size) {
  @media screen and (max-width: map-get($breakpoints, $size)) {
    @content;
  }
}
@mixin nudger($direction, $nudge) {
  @if $direction == 'horizontal' {
    width: #{$nudge}rem;
  }
  @if $direction == 'vertical' {
    height: #{$nudge}rem;
  }
}

// Vertical alignment helper
@mixin alignmentHelper() {
  // Add a char to align the component
  &::before {
    content: '\200C'; // zero-width non-joiner
    @include alignmentHelperClass();
  }
}
@mixin alignmentHelperClass() {
  display: inline-block;
  width: 0;
  height: 0;

  font-size: var(--font-size-small);

  speak: none;
}

// Screen reader only helper
@mixin srOnly() {
  user-select: none;
  -webkit-touch-callout: none; // iOS
  pointer-events: none;

  position: absolute;

  clip: rect(1px, 1px, 1px, 1px);
  overflow: hidden;
  white-space: nowrap; // NB: so NVDA not splits up text in multi line

  height: 1px;
  width: 1px; // splits up in multiline, unless we have nowrap
  margin: -1px 0 0 -1px;
  padding: 0;

  border: 0;
}

@mixin srOnlyInline() {
  @include srOnly();

  // since we do use it also as positioning helper
  // we then do not want it absolute positioned
  position: static;
  display: inline-block;

  visibility: visible; // should not be hidden!
  opacity: 0;

  font-size: 0;
  line-height: 0;
}

// This "negates" the srOnly() helper; useful if
// we want the default state of something to be
// for screen readers only, but we want to force
// it to show for sighted users under more specific
// conditions.
@mixin notSrOnly() {
  position: initial;
  height: auto;
  width: auto;

  overflow: auto;
  opacity: 1;

  font-size: medium;
  line-height: normal;
  white-space: normal;
}

@function str-replace($string, $search, $replace: '') {
  $index: str-index($string, $search);

  @if $index {
    @return str-slice($string, 1, $index - 1) + $replace +
      str-replace(
        str-slice($string, $index + str-length($search)),
        $search,
        $replace
      );
  }

  @return $string;
}

@mixin formLabelWrap() {
  @media screen and (max-width: 40em) {
    flex-wrap: wrap;
    & > .dnb-form-label {
      margin-bottom: 0.5rem;
      margin-top: 0.5rem;
    }
  }
}
