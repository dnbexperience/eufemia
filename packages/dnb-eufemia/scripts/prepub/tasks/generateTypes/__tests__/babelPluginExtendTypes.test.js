/**
 * Scripts test
 *
 */

import { transformFileAsync } from '@babel/core'
import nodePath from 'path'
import { babelPluginConfigDefaults } from '../babelPluginConfigDefaults'
import { babelPluginExtendTypes } from '../babelPluginExtendTypes'

describe('babelPluginExtendTypes', () => {
  it('has to match snapshot', async () => {
    const targetFile = nodePath.resolve(
      __dirname,
      '__mocks__/PrimaryComponent.d.ts'
    )
    const file = nodePath.resolve(
      __dirname,
      '__mocks__/PrimaryComponent.js'
    )

    const { code } = await transformFileAsync(targetFile, {
      plugins: [
        ['@babel/plugin-syntax-typescript', { isTSX: true }],
        [
          babelPluginExtendTypes,
          {
            file,
          },
        ],
      ],
      ...babelPluginConfigDefaults,
    })

    expect(code).toMatchSnapshot()
  })
})
