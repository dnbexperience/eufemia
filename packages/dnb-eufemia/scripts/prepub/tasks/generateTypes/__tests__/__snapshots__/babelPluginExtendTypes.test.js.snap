// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babelPluginExtendTypes has to match snapshot 1`] = `
"import * as React from 'react';
import SecondaryComponent, { secondaryDefaultProps, secondaryPropTypes } from './SecondaryComponent';
export interface PrimaryComponentSpacing {
  top?: string | number | boolean;
  right?: string | number | boolean;
  bottom?: string | number | boolean;
  left?: string | number | boolean;
}
export type PrimaryComponentFoo = string | number | boolean;
export type PrimaryComponentTop = string | number | boolean;
export type PrimaryComponentRight = string | number | boolean;
export type PrimaryComponentBottom = string | number | boolean;
export type PrimaryComponentLeft = string | number | boolean;
export type PrimaryComponentSecondary = string | number | boolean;
export type PrimaryComponentSecondaryFoo = string | number | boolean;
export interface PrimaryComponentSecondarySpacing {
  secondary?: string | number | boolean;
}
export interface PrimaryComponentProps extends React.HTMLProps<HTMLElement> {
  boolean?: boolean;
  number?: number;
  spacing?: PrimaryComponentSpacing;
  foo?: PrimaryComponentFoo;
  top?: PrimaryComponentTop;
  right?: PrimaryComponentRight;
  bottom?: PrimaryComponentBottom;
  left?: PrimaryComponentLeft;
  secondary?: PrimaryComponentSecondary;
  secondary_foo?: PrimaryComponentSecondaryFoo;
  secondary_spacing?: PrimaryComponentSecondarySpacing;
  children?: React.ReactNode;
}
export default class PrimaryComponent extends React.Component<PrimaryComponentProps, any> {
  static defaultProps: object;
  static Secondary = SecondaryComponent;
  render(): JSX.Element;
}"
`;
